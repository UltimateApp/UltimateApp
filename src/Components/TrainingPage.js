import React from 'react';
import { StyleSheet, View, Text, TouchableOpacity, Image } from 'react-native';
import GestureRecognizer, { swipeDirections } from 'react-native-swipe-gestures';

import DrillList from './shared/DrillList';
import Button from './shared/Button';
import theme from '../styles/theme.style';
import { swipeConfig } from '../styles/config';
import iconPlayers from '../../assets/ic_players.png';
import iconClock from '../../assets/ic_clock.png';
import { convertMinsToTime } from '../utils/time';

export function getTrainingDuration(training) {
  return training.drills.reduce((total, drill) => total + drill.durationInMinutes, 0);
}

export function getTrainingMinimalPlayersNumber(training) {
  const minimalPlayersNumberList = training.drills.map(({ minimalPlayersNumber }) => minimalPlayersNumber);
  return Math.max(...minimalPlayersNumberList);
}

export const TrainingPage = props => {
  const { navigation, route } = props;
  const { training, program } = route.params;

  const onDrillPress = drill => navigation.navigate('DrillPageMinimal', { drill, training, program });
  const goToFirstDrill = () =>
    navigation.navigate('DrillPageMinimal', { drill: training.drills[0], training, program });

  const currentTrainingIndex = program
    ? program.trainings.findIndex(programTraining => programTraining.id === training.id)
    : null;
  const isFirstTraining = program ? currentTrainingIndex === 0 : true;
  const isLastTraining = program ? currentTrainingIndex === program.trainings.length - 1 : true;

  const onNextPress = () => {
    if (!isLastTraining) {
      navigation.navigate('TrainingPage', { training: program.trainings[currentTrainingIndex + 1], program });
    }
  };
  const onPrevPress = () => {
    if (!isFirstTraining) {
      navigation.navigate('TrainingPage', { training: program.trainings[currentTrainingIndex - 1], program });
    }
  };
  const programNavigation = () => {
    return (
      <View style={styles.programNavigation}>
        {!isFirstTraining && (
          <TouchableOpacity style={styles.btnPrevNext} onPress={onPrevPress}>
            <Text style={styles.btnPrevNextContent}>{'<'}</Text>
          </TouchableOpacity>
        )}
        <Text style={styles.title}>{training.title}</Text>
        {program.trainings.length > 1 && (
          <Text style={styles.subtitle}>
            {' '}
            ({currentTrainingIndex + 1}/{program.trainings.length}){' '}
          </Text>
        )}
        {!isLastTraining && (
          <TouchableOpacity style={styles.btnPrevNext} onPress={onNextPress}>
            <Text style={styles.btnPrevNextContent}>{'>'}</Text>
          </TouchableOpacity>
        )}
      </View>
    );
  };

  const header = () => (
    <View style={styles.overview}>
      {program && programNavigation()}
      <View style={styles.infos}>
        <View style={styles.info}>
          <Image style={styles.infoIcon} source={iconPlayers} />
          <Text style={styles.infoValue}>{getTrainingMinimalPlayersNumber(training)}+</Text>
        </View>
        <View style={styles.info}>
          <Image style={styles.infoIcon} source={iconClock} />
          <Text style={styles.infoValue}>{convertMinsToTime(getTrainingDuration(training))}</Text>
        </View>
      </View>
      <Text style={styles.descriptionText}>{training.description}</Text>
    </View>
  );

  return (
    <GestureRecognizer style={{ flex: 1 }} onSwipeLeft={onNextPress} onSwipeRight={onPrevPress} config={swipeConfig}>
      <DrillList
        style={styles.trainingPage}
        ListHeaderComponent={header}
        ListFooterComponent={<View style={{ paddingBottom: 30 }} />}
        ItemComponentStyle={styles.list}
        navigation={navigation}
        drillsToDisplay={training.drills}
        onDrillPress={onDrillPress}
      />
      <View style={styles.footer}>
        <Button onPress={goToFirstDrill} text="Start training" />
      </View>
    </GestureRecognizer>
  );
};

export default TrainingPage;

const styles = StyleSheet.create({
  trainingPage: {
    backgroundColor: theme.BACKGROUND_COLOR_LIGHT,
    height: '100%',
    flex: 1,
  },
  trainingPageContent: {
    paddingBottom: 50,
  },
  overview: {
    paddingLeft: 20,
    paddingBottom: 20,
    marginVertical: 20,
    borderBottomWidth: 2,
    borderBottomColor: theme.COLOR_SECONDARY_LIGHT,
  },
  programNavigation: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
  },
  title: {
    fontSize: theme.FONT_SIZE_LARGE,
    fontWeight: 'bold',
  },
  subtitle: {
    fontSize: theme.FONT_SIZE_LARGE,
  },
  btnPrevNext: {
    padding: 10,
  },
  btnPrevNextContent: {
    textAlign: 'center',
    color: theme.COLOR_PRIMARY,
    fontSize: theme.FONT_SIZE_LARGE,
  },
  descriptionText: {
    color: theme.COLOR_SECONDARY,
    fontSize: theme.FONT_SIZE_MEDIUM,
  },
  infos: {
    flexDirection: 'row',
    justifyContent: 'space-evenly',
  },
  info: {
    alignItems: 'center',
  },
  infoIcon: {
    width: 20,
    height: 20,
  },
  infoValue: {
    color: theme.COLOR_SECONDARY,
    fontSize: theme.FONT_SIZE_MEDIUM,
  },
  list: {
    paddingLeft: 20,
  },
  footer: {
    position: 'absolute',
    paddingBottom: 20,
    paddingTop: 5,
    bottom: 0,
    backgroundColor: 'white',
    width: '100%',
    alignItems: 'center',
  },
});
