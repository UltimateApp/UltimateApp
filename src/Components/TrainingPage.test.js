import React from 'react';
import renderer from 'react-test-renderer';
import { render, fireEvent, cleanup } from 'react-native-testing-library';
import fixtures from '../Fixtures/TestFixtures';

import TrainingPage from './TrainingPage';

afterEach(cleanup);

describe('<TrainingPage />', () => {
  const navigation = { navigate: jest.fn() };
  const program = fixtures.programs[0];
  const training = program.trainings[0];
  let route = {
    params: {
      training,
      program,
    },
  };

  it('renders correctly with a training outside any program', () => {
    const route = {
      params: {
        training,
      },
    };
    const tree = renderer.create(<TrainingPage route={route} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with a training within a program', () => {
    const tree = renderer.create(<TrainingPage route={route} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('links to previous training within program', async () => {
    route = {
      params: {
        training: program.trainings[1],
        program,
      },
    };

    const { queryByText } = render(<TrainingPage navigation={navigation} route={route} />);

    expect(queryByText('<')).toBeDefined();
    expect(queryByText('>')).toBeNull(); // Not displayed on last training

    await fireEvent.press(queryByText('<'));

    expect(navigation.navigate).toBeCalledWith('TrainingPage', { training: program.trainings[0], program });
  });

  it('links to next training within program', async () => {
    route = {
      params: {
        training: program.trainings[0],
        program,
      },
    };

    const { queryByText } = render(<TrainingPage navigation={navigation} route={route} />);

    expect(queryByText('<')).toBeNull();
    expect(queryByText('>')).toBeDefined();

    await fireEvent.press(queryByText('>'));

    expect(navigation.navigate).toBeCalledWith('TrainingPage', { training: program.trainings[1], program });
  });

  it('links to drill page', async () => {
    const trainingFirstDrill = training.drills[0];

    const { queryByText } = render(<TrainingPage navigation={navigation} route={route} />);

    await fireEvent.press(queryByText(trainingFirstDrill.title));

    expect(navigation.navigate).toBeCalledWith('DrillPageMinimal', { drill: trainingFirstDrill, training, program });
  });

  it('starts training links to first drill', async () => {
    const { queryByText } = render(<TrainingPage navigation={navigation} route={route} />);

    await fireEvent.press(queryByText('Start training'));

    expect(navigation.navigate).toBeCalledWith('DrillPageMinimal', { drill: training.drills[0], training, program });
  });
});
