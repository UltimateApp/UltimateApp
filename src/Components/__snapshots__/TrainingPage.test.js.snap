// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TrainingPage /> renders correctly with a training 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#fff",
      "height": "100%",
      "paddingLeft": 20,
      "paddingTop": 10,
    }
  }
>
  <Text
    style={
      Object {
        "color": "#A6A6A6",
        "fontSize": 16,
        "padding": 20,
        "textAlign": "center",
      }
    }
  >
    Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec scelerisque dignissim felis, at tristique mi efficitur a. Nulla pellentesque odio nunc, ut sodales ex pulvinar at. In hac habitasse platea dictumst. Pellentesque sodales nisl lorem, ac lacinia nisl lacinia eu. Proin lacinia viverra mauris, et pharetra ipsum tempus eget. Ut maximus sapien in hendrerit ultricies. Maecenas vel diam tincidunt, consequat nunc a, mattis eros. Curabitur a eros a nulla malesuada gravida. Donec massa quam, suscipit eget efficitur et, euismod sed nulla. Morbi mattis et magna a aliquam. Fusce pellentesque vel erat eget
  </Text>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-evenly",
        "padding": 10,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#232b2b",
          "fontSize": 16,
          "fontWeight": "bold",
          "paddingHorizontal": 30,
        }
      }
    >
      Players:
    </Text>
    <Text
      style={
        Object {
          "color": "#A6A6A6",
          "fontSize": 16,
          "paddingHorizontal": 30,
        }
      }
    >
      2
      +
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-evenly",
        "padding": 10,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#232b2b",
          "fontSize": 16,
          "fontWeight": "bold",
          "paddingHorizontal": 30,
        }
      }
    >
      Goals:
    </Text>
    <Text
      style={
        Object {
          "color": "#A6A6A6",
          "fontSize": 16,
          "paddingHorizontal": 30,
        }
      }
    >
      Legs, Defense
    </Text>
  </View>
  <View>
    <RCTScrollView
      data={
        Array [
          Object {
            "description": "Description of the drill",
            "durationInMinutes": 10,
            "equipment": "Equipment needs for the drill",
            "equipmentLabel": "Full equipment",
            "goals": Array [
              "Legs",
            ],
            "id": 2,
            "image": "http://www.liberte-fitness.fr/sites/default/files/styles/slider/public/news/tone.jpg?itok=505bme2a",
            "intensity": "High",
            "level": "Intermediate",
            "minimalPlayersNumber": 1,
            "seasonTiming": "In-season",
            "source": "Author",
            "steps": Array [
              Object {
                "subtitle": "This is a subtitle",
                "title": "Warmup",
              },
              Object {
                "subtitle": "This is a subtitle",
                "title": "Do the drill",
              },
              Object {
                "subtitle": "This is a subtitle",
                "title": "Last step",
              },
            ],
            "title": "Fitness Drill Title",
            "type": "fitness",
            "video": "https://www.youtube.com/embed/JkVHrA5o23o",
          },
          Object {
            "animation": Object {
              "ids": Array [
                "triangle",
                "triangle",
                "triangle",
                "triangle",
                "triangle",
                "triangle",
                "triangle",
                "triangle",
                "offense",
                "offense",
                "offense",
                "offense",
                "offense",
                "disc",
              ],
              "positions": Array [
                Array [
                  Array [
                    Array [
                      0.3,
                      0.3,
                    ],
                  ],
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                ],
                Array [
                  Array [
                    Array [
                      0.3,
                      0.6,
                    ],
                  ],
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                ],
                Array [
                  Array [
                    Array [
                      0.6,
                      0.3,
                    ],
                  ],
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                ],
                Array [
                  Array [
                    Array [
                      0.6,
                      0.6,
                    ],
                  ],
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                ],
                Array [
                  Array [
                    Array [
                      0.45,
                      0.2,
                    ],
                  ],
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                ],
                Array [
                  Array [
                    Array [
                      0.2,
                      0.45,
                    ],
                  ],
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                ],
                Array [
                  Array [
                    Array [
                      0.45,
                      0.7,
                    ],
                  ],
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                ],
                Array [
                  Array [
                    Array [
                      0.7,
                      0.45,
                    ],
                  ],
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                ],
                Array [
                  Array [
                    Array [
                      0.3,
                      0.3,
                    ],
                  ],
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  Array [
                    Array [
                      0.45,
                      0.2,
                    ],
                  ],
                ],
                Array [
                  Array [
                    Array [
                      0.16,
                      0.3,
                    ],
                  ],
                  Array [
                    Array [
                      0.45,
                      0.2,
                    ],
                  ],
                  Array [
                    Array [
                      0.6,
                      0.3,
                    ],
                  ],
                  undefined,
                  undefined,
                  undefined,
                ],
                Array [
                  Array [
                    Array [
                      0.6,
                      0.3,
                    ],
                  ],
                  undefined,
                  Array [
                    Array [
                      0.7,
                      0.45,
                    ],
                  ],
                  Array [
                    Array [
                      0.6,
                      0.6,
                    ],
                  ],
                  undefined,
                  undefined,
                ],
                Array [
                  Array [
                    Array [
                      0.6,
                      0.6,
                    ],
                  ],
                  undefined,
                  undefined,
                  Array [
                    Array [
                      0.45,
                      0.7,
                    ],
                  ],
                  Array [
                    Array [
                      0.3,
                      0.6,
                    ],
                  ],
                  undefined,
                ],
                Array [
                  Array [
                    Array [
                      0.3,
                      0.6,
                    ],
                  ],
                  undefined,
                  undefined,
                  undefined,
                  Array [
                    Array [
                      0.2,
                      0.45,
                    ],
                  ],
                  Array [
                    Array [
                      0.3,
                      0.3,
                    ],
                  ],
                ],
                Array [
                  Array [
                    Array [
                      0.36,
                      0.36,
                    ],
                  ],
                  undefined,
                  Array [
                    Array [
                      0.58,
                      0.36,
                    ],
                  ],
                  Array [
                    Array [
                      0.58,
                      0.58,
                    ],
                  ],
                  Array [
                    Array [
                      0.36,
                      0.58,
                    ],
                  ],
                  Array [
                    Array [
                      0.36,
                      0.36,
                    ],
                  ],
                ],
              ],
              "texts": Array [
                "",
                "",
                "",
                "",
                "",
                "",
                "",
                "",
                "1",
                "2",
                "3",
                "4",
                "5",
                "",
              ],
            },
            "description": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec scelerisque dignissim felis, at tristique mi efficitur a. Nulla pellentesque odio nunc, ut sodales ex pulvinar at. In hac habitasse platea dictumst. Pellentesque sodales nisl lorem, ac lacinia nisl lacinia eu. Proin lacinia viverra mauris, et pharetra ipsum tempus eget. Ut maximus sapien in hendrerit ultricies. Maecenas vel diam tincidunt, consequat nunc a, mattis eros. Curabitur a eros a nulla malesuada gravida. Donec massa quam, suscipit eget efficitur et, euismod sed nulla. Morbi mattis et magna a aliquam. Fusce pellentesque vel erat eget",
            "durationInMinutes": 30,
            "equipment": "Equipment needs for the drill",
            "equipmentLabel": "",
            "goals": Array [
              "Defense",
            ],
            "id": 3,
            "image": "https://i1.wp.com/www.newdelhitimes.com/wp-content/uploads/2019/07/AP19204604544330.jpg?w=1024&ssl=1",
            "intensity": "Low",
            "level": "Beginner",
            "minimalPlayersNumber": 2,
            "seasonTiming": "",
            "source": "Author",
            "steps": Array [],
            "title": "Frisbee Drill Title",
            "type": "technical",
          },
        ]
      }
      disableVirtualization={false}
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      numColumns={1}
      onContentSizeChange={[Function]}
      onEndReachedThreshold={2}
      onLayout={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      updateCellsBatchingPeriod={50}
      viewabilityConfigCallbackPairs={Array []}
      windowSize={21}
    >
      <View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            accessible={true}
            focusable={true}
            isTVSelectable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "flexDirection": "row",
                "height": 80,
                "marginBottom": 20,
                "opacity": 1,
              }
            }
          >
            <View
              accessibilityIgnoresInvertColors={true}
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "rgb(0,0,0)",
                  "borderRadius": 5,
                  "height": 80,
                  "justifyContent": "center",
                  "marginRight": 10,
                  "padding": 10,
                  "width": 80,
                }
              }
            >
              <Image
                source={
                  Object {
                    "uri": "http://www.liberte-fitness.fr/sites/default/files/styles/slider/public/news/tone.jpg?itok=505bme2a",
                  }
                }
                style={
                  Array [
                    Object {
                      "bottom": 0,
                      "left": 0,
                      "position": "absolute",
                      "right": 0,
                      "top": 0,
                    },
                    Object {
                      "height": 80,
                      "width": 80,
                    },
                    Object {
                      "borderRadius": 5,
                      "opacity": 0.5,
                    },
                  ]
                }
              />
              <Text
                style={
                  Object {
                    "color": "#e5e5e5",
                    "fontSize": 20,
                    "textAlign": "center",
                  }
                }
              >
                10
                +
              </Text>
              <Text
                style={
                  Object {
                    "color": "#e5e5e5",
                    "fontSize": 14,
                    "textAlign": "center",
                  }
                }
              >
                min.
              </Text>
            </View>
            <View
              style={
                Object {
                  "padding": 5,
                  "paddingBottom": 10,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#A6A6A6",
                    "flex": 2,
                    "fontSize": 14,
                  }
                }
              >
                Author
              </Text>
              <Text
                style={
                  Object {
                    "flex": 3,
                    "flexWrap": "wrap",
                    "fontSize": 20,
                    "fontWeight": "bold",
                  }
                }
              >
                Fitness Drill Title
              </Text>
              <Text
                style={
                  Object {
                    "color": "#A6A6A6",
                    "flex": 2,
                    "fontSize": 14,
                  }
                }
              >
                Legs
              </Text>
            </View>
          </View>
        </View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            accessible={true}
            focusable={true}
            isTVSelectable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "flexDirection": "row",
                "height": 80,
                "marginBottom": 20,
                "opacity": 1,
              }
            }
          >
            <View
              accessibilityIgnoresInvertColors={true}
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "rgb(0,0,0)",
                  "borderRadius": 5,
                  "height": 80,
                  "justifyContent": "center",
                  "marginRight": 10,
                  "padding": 10,
                  "width": 80,
                }
              }
            >
              <Image
                source={
                  Object {
                    "uri": "https://i1.wp.com/www.newdelhitimes.com/wp-content/uploads/2019/07/AP19204604544330.jpg?w=1024&ssl=1",
                  }
                }
                style={
                  Array [
                    Object {
                      "bottom": 0,
                      "left": 0,
                      "position": "absolute",
                      "right": 0,
                      "top": 0,
                    },
                    Object {
                      "height": 80,
                      "width": 80,
                    },
                    Object {
                      "borderRadius": 5,
                      "opacity": 0.5,
                    },
                  ]
                }
              />
              <Text
                style={
                  Object {
                    "color": "#e5e5e5",
                    "fontSize": 20,
                    "textAlign": "center",
                  }
                }
              >
                2
                +
              </Text>
              <Text
                style={
                  Object {
                    "color": "#e5e5e5",
                    "fontSize": 14,
                    "textAlign": "center",
                  }
                }
              >
                players
              </Text>
            </View>
            <View
              style={
                Object {
                  "padding": 5,
                  "paddingBottom": 10,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#A6A6A6",
                    "flex": 2,
                    "fontSize": 14,
                  }
                }
              >
                Author
              </Text>
              <Text
                style={
                  Object {
                    "flex": 3,
                    "flexWrap": "wrap",
                    "fontSize": 20,
                    "fontWeight": "bold",
                  }
                }
              >
                Frisbee Drill Title
              </Text>
              <Text
                style={
                  Object {
                    "color": "#A6A6A6",
                    "flex": 2,
                    "fontSize": 14,
                  }
                }
              >
                Defense
              </Text>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
</View>
`;
